src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs

```diff
@@ -177,11 +177,10 @@ internal ConcurrentDictionary(int concurrencyLevel, int capacity, bool growLockA
             }
             capacity = HashHelpers.GetPrime(capacity);
 
-            var locks = new object[concurrencyLevel];
-            locks[0] = locks; // reuse array as the first lock object just to avoid an additional allocation
-            for (int i = 1; i < locks.Length; i++)
+            var locks = new Lock[concurrencyLevel];
+            for (int i = 0; i < locks.Length; i++)
             {
-                locks[i] = new object();
+                locks[i] = new Lock();
```

Comment:
This is now allocating an extra object. Is it worth it?

Reply:
The microbenchmarks from the performance repo that use ConcurrentDictionary were tested with this change and got the following results:
```
// * Summary *

BenchmarkDotNet v0.13.13-nightly.20240311.145, Windows 11 (10.0.22631.3737/23H2/2023Update/SunValley3) (Hyper-V)
AMD EPYC 7763, 1 CPU, 16 logical and 8 physical cores
.NET SDK 9.0.100-preview.2.24157.14
  [Host]     : .NET 9.0.0 (9.0.24.12805), X64 RyuJIT AVX2
  Job-DNQFPU : .NET 9.0.0 (42.42.42.42424), X64 RyuJIT AVX2
  Job-OSLZLB : .NET 9.0.0 (42.42.42.42424), X64 RyuJIT AVX2

PowerPlanMode=00000000-0000-0000-0000-000000000000  IterationTime=250ms  MaxIterationCount=20
MinIterationCount=15  WarmupCount=1

| Type                             | Method               | Job        | Toolchain                | Count | Size | Mean        | Error     | StdDev    | Median      | Min         | Max         | Ratio | RatioSD | Gen0   | Gen1   | Allocated | Alloc Ratio |
|--------------------------------- |--------------------- |----------- |------------------------- |------ |----- |------------:|----------:|----------:|------------:|------------:|------------:|------:|--------:|-------:|-------:|----------:|------------:|
| CtorDefaultSize<Int32>           | ConcurrentDictionary | Job-DNQFPU | \CD-Baseline\corerun.exe | ?     | ?    |    144.0 ns |   1.01 ns |   0.85 ns |    143.8 ns |    142.6 ns |    145.4 ns |  1.00 |    0.01 | 0.0597 |      - |    1008 B |        1.00 |
| CtorDefaultSize<Int32>           | ConcurrentDictionary | Job-OSLZLB | \CD-Changed\corerun.exe  | ?     | ?    |    141.2 ns |   1.24 ns |   1.10 ns |    141.2 ns |    139.7 ns |    143.7 ns |  0.98 |    0.01 | 0.0600 |      - |    1008 B |        1.00 |
|                                  |                      |            |                          |       |      |             |           |           |             |             |             |       |         |        |        |           |             |
| CtorDefaultSize<String>          | ConcurrentDictionary | Job-DNQFPU | \CD-Baseline\corerun.exe | ?     | ?    |    157.3 ns |   1.55 ns |   1.37 ns |    157.3 ns |    154.5 ns |    159.7 ns |  1.00 |    0.01 | 0.0598 |      - |    1008 B |        1.00 |
| CtorDefaultSize<String>          | ConcurrentDictionary | Job-OSLZLB | \CD-Changed\corerun.exe  | ?     | ?    |    156.4 ns |   1.54 ns |   1.37 ns |    156.5 ns |    153.5 ns |    158.2 ns |  0.99 |    0.01 | 0.0602 |      - |    1008 B |        1.00 |
|                                  |                      |            |                          |       |      |             |           |           |             |             |             |       |         |        |        |           |             |
| TryAddDefaultSize<Int32>         | ConcurrentDictionary | Job-DNQFPU | \CD-Baseline\corerun.exe | 512   | ?    | 35,922.3 ns | 330.25 ns | 275.77 ns | 35,931.4 ns | 35,599.5 ns | 36,480.5 ns |  1.00 |    0.01 | 5.7078 | 0.9989 |   96672 B |        1.00 |
| TryAddDefaultSize<Int32>         | ConcurrentDictionary | Job-OSLZLB | \CD-Changed\corerun.exe  | 512   | ?    | 36,358.9 ns | 251.52 ns | 222.96 ns | 36,375.8 ns | 36,001.9 ns | 36,626.5 ns |  1.01 |    0.01 | 5.7604 | 1.0081 |   96672 B |        1.00 |
|                                  |                      |            |                          |       |      |             |           |           |             |             |             |       |         |        |        |           |             |
| TryAddDefaultSize<String>        | ConcurrentDictionary | Job-DNQFPU | \CD-Baseline\corerun.exe | 512   | ?    | 47,174.1 ns | 275.94 ns | 244.61 ns | 47,233.1 ns | 46,760.3 ns | 47,559.5 ns |  1.00 |    0.01 | 6.2123 | 1.5060 |  105736 B |        1.00 |
| TryAddDefaultSize<String>        | ConcurrentDictionary | Job-OSLZLB | \CD-Changed\corerun.exe  | 512   | ?    | 47,186.2 ns | 272.78 ns | 241.81 ns | 47,162.9 ns | 46,805.9 ns | 47,618.6 ns |  1.00 |    0.01 | 6.2123 | 1.5060 |  105736 B |        1.00 |
|                                  |                      |            |                          |       |      |             |           |           |             |             |             |       |         |        |        |           |             |
| TryAddGiventSize<Int32>          | ConcurrentDictionary | Job-DNQFPU | \CD-Baseline\corerun.exe | 512   | ?    | 23,323.2 ns | 146.40 ns | 129.78 ns | 23,332.9 ns | 23,132.7 ns | 23,589.4 ns |  1.00 |    0.01 | 3.1575 | 0.4643 |   53192 B |        1.00 |
| TryAddGiventSize<Int32>          | ConcurrentDictionary | Job-OSLZLB | \CD-Changed\corerun.exe  | 512   | ?    | 23,383.0 ns | 235.54 ns | 196.68 ns | 23,426.3 ns | 23,057.4 ns | 23,676.4 ns |  1.00 |    0.01 | 3.1489 | 0.4771 |   53192 B |        1.00 |
|                                  |                      |            |                          |       |      |             |           |           |             |             |             |       |         |        |        |           |             |
| TryAddGiventSize<String>         | ConcurrentDictionary | Job-DNQFPU | \CD-Baseline\corerun.exe | 512   | ?    | 33,580.3 ns | 211.27 ns | 187.28 ns | 33,562.8 ns | 33,322.6 ns | 33,907.8 ns |  1.00 |    0.01 | 3.5828 | 0.6635 |   60624 B |        1.00 |
| TryAddGiventSize<String>         | ConcurrentDictionary | Job-OSLZLB | \CD-Changed\corerun.exe  | 512   | ?    | 33,639.2 ns | 229.71 ns | 191.82 ns | 33,619.2 ns | 33,328.7 ns | 33,962.8 ns |  1.00 |    0.01 | 3.6212 | 0.6706 |   60624 B |        1.00 |
|                                  |                      |            |                          |       |      |             |           |           |             |             |             |       |         |        |        |           |             |
| ContainsKeyFalse<Int32, Int32>   | ConcurrentDictionary | Job-DNQFPU | \CD-Baseline\corerun.exe | ?     | 512  |  2,137.6 ns |   7.89 ns |   6.59 ns |  2,136.2 ns |  2,128.4 ns |  2,149.9 ns |  1.00 |    0.00 |      - |      - |         - |          NA |
| ContainsKeyFalse<Int32, Int32>   | ConcurrentDictionary | Job-OSLZLB | \CD-Changed\corerun.exe  | ?     | 512  |  2,136.8 ns |   2.58 ns |   2.41 ns |  2,136.3 ns |  2,132.6 ns |  2,141.2 ns |  1.00 |    0.00 |      - |      - |         - |          NA |
|                                  |                      |            |                          |       |      |             |           |           |             |             |             |       |         |        |        |           |             |
| ContainsKeyFalse<String, String> | ConcurrentDictionary | Job-DNQFPU | \CD-Baseline\corerun.exe | ?     | 512  |  6,686.9 ns |  19.38 ns |  17.18 ns |  6,693.7 ns |  6,641.8 ns |  6,701.3 ns |  1.00 |    0.00 |      - |      - |         - |          NA |
| ContainsKeyFalse<String, String> | ConcurrentDictionary | Job-OSLZLB | \CD-Changed\corerun.exe  | ?     | 512  |  6,709.3 ns |  11.29 ns |   9.43 ns |  6,708.8 ns |  6,697.6 ns |  6,726.2 ns |  1.00 |    0.00 |      - |      - |         - |          NA |
|                                  |                      |            |                          |       |      |             |           |           |             |             |             |       |         |        |        |           |             |
| ContainsKeyTrue<Int32, Int32>    | ConcurrentDictionary | Job-DNQFPU | \CD-Baseline\corerun.exe | ?     | 512  |  2,327.9 ns |   4.52 ns |   4.23 ns |  2,327.0 ns |  2,320.0 ns |  2,335.9 ns |  1.00 |    0.00 |      - |      - |         - |          NA |
| ContainsKeyTrue<Int32, Int32>    | ConcurrentDictionary | Job-OSLZLB | \CD-Changed\corerun.exe  | ?     | 512  |  2,327.1 ns |   5.56 ns |   4.93 ns |  2,326.1 ns |  2,317.9 ns |  2,334.5 ns |  1.00 |    0.00 |      - |      - |         - |          NA |
|                                  |                      |            |                          |       |      |             |           |           |             |             |             |       |         |        |        |           |             |
| ContainsKeyTrue<String, String>  | ConcurrentDictionary | Job-DNQFPU | \CD-Baseline\corerun.exe | ?     | 512  |  9,101.7 ns |  32.03 ns |  28.40 ns |  9,099.0 ns |  9,056.0 ns |  9,162.4 ns |  1.00 |    0.00 |      - |      - |         - |          NA |
| ContainsKeyTrue<String, String>  | ConcurrentDictionary | Job-OSLZLB | \CD-Changed\corerun.exe  | ?     | 512  |  9,071.8 ns |  28.46 ns |  23.77 ns |  9,074.3 ns |  9,006.6 ns |  9,105.0 ns |  1.00 |    0.00 |      - |      - |         - |          NA |
|                                  |                      |            |                          |       |      |             |           |           |             |             |             |       |         |        |        |           |             |
| TryGetValueFalse<Int32, Int32>   | ConcurrentDictionary | Job-DNQFPU | \CD-Baseline\corerun.exe | ?     | 512  |    804.6 ns |   3.61 ns |   3.01 ns |    803.8 ns |    801.1 ns |    812.2 ns |  1.00 |    0.01 |      - |      - |         - |          NA |
| TryGetValueFalse<Int32, Int32>   | ConcurrentDictionary | Job-OSLZLB | \CD-Changed\corerun.exe  | ?     | 512  |    803.1 ns |   2.26 ns |   1.88 ns |    802.7 ns |    800.7 ns |    807.1 ns |  1.00 |    0.00 |      - |      - |         - |          NA |
|                                  |                      |            |                          |       |      |             |           |           |             |             |             |       |         |        |        |           |             |
| TryGetValueFalse<String, String> | ConcurrentDictionary | Job-DNQFPU | \CD-Baseline\corerun.exe | ?     | 512  |  5,393.8 ns |  23.89 ns |  22.35 ns |  5,394.1 ns |  5,343.0 ns |  5,425.8 ns |  1.00 |    0.01 |      - |      - |         - |          NA |
| TryGetValueFalse<String, String> | ConcurrentDictionary | Job-OSLZLB | \CD-Changed\corerun.exe  | ?     | 512  |  5,393.6 ns |  13.51 ns |  11.28 ns |  5,390.3 ns |  5,381.0 ns |  5,419.2 ns |  1.00 |    0.00 |      - |      - |         - |          NA |
|                                  |                      |            |                          |       |      |             |           |           |             |             |             |       |         |        |        |           |             |
| TryGetValueTrue<Int32, Int32>    | ConcurrentDictionary | Job-DNQFPU | \CD-Baseline\corerun.exe | ?     | 512  |  1,020.8 ns |   4.05 ns |   3.59 ns |  1,019.8 ns |  1,016.9 ns |  1,029.2 ns |  1.00 |    0.00 |      - |      - |         - |          NA |
| TryGetValueTrue<Int32, Int32>    | ConcurrentDictionary | Job-OSLZLB | \CD-Changed\corerun.exe  | ?     | 512  |  1,021.2 ns |   4.37 ns |   4.09 ns |  1,021.3 ns |  1,014.6 ns |  1,030.4 ns |  1.00 |    0.01 |      - |      - |         - |          NA |
|                                  |                      |            |                          |       |      |             |           |           |             |             |             |       |         |        |        |           |             |
| TryGetValueTrue<String, String>  | ConcurrentDictionary | Job-DNQFPU | \CD-Baseline\corerun.exe | ?     | 512  |  6,742.8 ns |  16.08 ns |  13.43 ns |  6,739.4 ns |  6,720.0 ns |  6,761.2 ns |  1.00 |    0.00 |      - |      - |         - |          NA |
| TryGetValueTrue<String, String>  | ConcurrentDictionary | Job-OSLZLB | \CD-Changed\corerun.exe  | ?     | 512  |  6,739.1 ns |   9.89 ns |   8.77 ns |  6,739.2 ns |  6,726.4 ns |  6,760.2 ns |  1.00 |    0.00 |      - |      - |         - |          NA |
|                                  |                      |            |                          |       |      |             |           |           |             |             |             |       |         |        |        |           |             |
| CtorGivenSize<Int32>             | ConcurrentDictionary | Job-DNQFPU | \CD-Baseline\corerun.exe | ?     | 512  |    195.1 ns |   6.23 ns |   6.93 ns |    194.8 ns |    180.5 ns |    207.7 ns |  1.00 |    0.05 | 0.2652 | 0.0039 |    4448 B |        1.00 |
| CtorGivenSize<Int32>             | ConcurrentDictionary | Job-OSLZLB | \CD-Changed\corerun.exe  | ?     | 512  |    191.8 ns |   4.12 ns |   4.75 ns |    191.4 ns |    182.9 ns |    201.9 ns |  0.98 |    0.04 | 0.2655 | 0.0038 |    4448 B |        1.00 |
|                                  |                      |            |                          |       |      |             |           |           |             |             |             |       |         |        |        |           |             |
| CtorGivenSize<String>            | ConcurrentDictionary | Job-DNQFPU | \CD-Baseline\corerun.exe | ?     | 512  |    201.3 ns |   4.12 ns |   4.58 ns |    202.9 ns |    192.4 ns |    206.3 ns |  1.00 |    0.03 | 0.2654 | 0.0042 |    4448 B |        1.00 |
| CtorGivenSize<String>            | ConcurrentDictionary | Job-OSLZLB | \CD-Changed\corerun.exe  | ?     | 512  |    201.7 ns |   3.50 ns |   3.10 ns |    202.3 ns |    193.9 ns |    204.8 ns |  1.00 |    0.03 | 0.2653 | 0.0040 |    4448 B |        1.00 |
|                                  |                      |            |                          |       |      |             |           |           |             |             |             |       |         |        |        |           |             |
| AddGivenSize<Int32>              | ConcurrentDictionary | Job-DNQFPU | \CD-Baseline\corerun.exe | ?     | 512  | 24,521.7 ns | 220.84 ns | 184.41 ns | 24,491.2 ns | 24,228.4 ns | 24,849.8 ns |  1.00 |    0.01 | 3.1596 | 0.4937 |   53192 B |        1.00 |
| AddGivenSize<Int32>              | ConcurrentDictionary | Job-OSLZLB | \CD-Changed\corerun.exe  | ?     | 512  | 23,496.6 ns | 208.38 ns | 184.72 ns | 23,468.2 ns | 23,285.0 ns | 23,924.2 ns |  0.96 |    0.01 | 3.0922 | 0.4685 |   53192 B |        1.00 |
|                                  |                      |            |                          |       |      |             |           |           |             |             |             |       |         |        |        |           |             |
| AddGivenSize<String>             | ConcurrentDictionary | Job-DNQFPU | \CD-Baseline\corerun.exe | ?     | 512  | 33,668.7 ns | 442.19 ns | 413.63 ns | 33,704.4 ns | 33,024.6 ns | 34,637.8 ns |  1.00 |    0.02 | 3.5981 | 0.6663 |   60624 B |        1.00 |
| AddGivenSize<String>             | ConcurrentDictionary | Job-OSLZLB | \CD-Changed\corerun.exe  | ?     | 512  | 33,715.1 ns | 284.86 ns | 266.46 ns | 33,644.6 ns | 33,361.6 ns | 34,277.4 ns |  1.00 |    0.01 | 3.5022 | 0.6735 |   60624 B |        1.00 |
|                                  |                      |            |                          |       |      |             |           |           |             |             |             |       |         |        |        |           |             |
| CtorFromCollection<Int32>        | ConcurrentDictionary | Job-DNQFPU | \CD-Baseline\corerun.exe | ?     | 512  | 20,918.6 ns | 160.24 ns | 142.05 ns | 20,920.5 ns | 20,661.5 ns | 21,121.7 ns |  1.00 |    0.01 | 2.9801 | 0.4139 |   50960 B |        1.00 |
| CtorFromCollection<Int32>        | ConcurrentDictionary | Job-OSLZLB | \CD-Changed\corerun.exe  | ?     | 512  | 20,901.6 ns | 118.50 ns | 105.04 ns | 20,914.6 ns | 20,731.2 ns | 21,046.3 ns |  1.00 |    0.01 | 3.0201 | 0.4195 |   50960 B |        1.00 |
|                                  |                      |            |                          |       |      |             |           |           |             |             |             |       |         |        |        |           |             |
| CtorFromCollection<String>       | ConcurrentDictionary | Job-DNQFPU | \CD-Baseline\corerun.exe | ?     | 512  | 33,400.9 ns | 133.03 ns | 117.93 ns | 33,389.2 ns | 33,234.2 ns | 33,564.5 ns |  1.00 |    0.00 | 3.4648 | 0.5330 |   58568 B |        1.00 |
| CtorFromCollection<String>       | ConcurrentDictionary | Job-OSLZLB | \CD-Changed\corerun.exe  | ?     | 512  | 33,551.6 ns | 388.80 ns | 324.66 ns | 33,569.8 ns | 33,161.5 ns | 34,323.8 ns |  1.00 |    0.01 | 3.4797 | 0.5353 |   58568 B |        1.00 |
|                                  |                      |            |                          |       |      |             |           |           |             |             |             |       |         |        |        |           |             |
| CreateAddAndClear<Int32>         | ConcurrentDictionary | Job-DNQFPU | \CD-Baseline\corerun.exe | ?     | 512  | 41,438.0 ns | 654.23 ns | 579.96 ns | 41,264.7 ns | 40,867.1 ns | 42,646.3 ns |  1.00 |    0.02 | 5.9211 | 0.8224 |   99120 B |        1.00 |
| CreateAddAndClear<Int32>         | ConcurrentDictionary | Job-OSLZLB | \CD-Changed\corerun.exe  | ?     | 512  | 41,429.1 ns | 219.18 ns | 183.03 ns | 41,411.3 ns | 41,016.4 ns | 41,702.2 ns |  1.00 |    0.01 | 5.7870 | 0.8267 |   99120 B |        1.00 |
|                                  |                      |            |                          |       |      |             |           |           |             |             |             |       |         |        |        |           |             |
| CreateAddAndClear<String>        | ConcurrentDictionary | Job-DNQFPU | \CD-Baseline\corerun.exe | ?     | 512  | 52,551.9 ns | 369.41 ns | 308.47 ns | 52,611.9 ns | 51,946.2 ns | 53,077.2 ns |  1.00 |    0.01 | 6.3131 | 0.8418 |  108184 B |        1.00 |
| CreateAddAndClear<String>        | ConcurrentDictionary | Job-OSLZLB | \CD-Changed\corerun.exe  | ?     | 512  | 52,465.0 ns | 463.08 ns | 361.54 ns | 52,401.8 ns | 51,956.7 ns | 53,160.6 ns |  1.00 |    0.01 | 6.2709 | 0.8361 |  108184 B |        1.00 |
|                                  |                      |            |                          |       |      |             |           |           |             |             |             |       |         |        |        |           |             |
| IterateForEach<Int32>            | ConcurrentDictionary | Job-DNQFPU | \CD-Baseline\corerun.exe | ?     | 512  |  6,503.5 ns |  18.36 ns |  16.28 ns |  6,500.9 ns |  6,484.9 ns |  6,535.1 ns |  1.00 |    0.00 |      - |      - |      56 B |        1.00 |
| IterateForEach<Int32>            | ConcurrentDictionary | Job-OSLZLB | \CD-Changed\corerun.exe  | ?     | 512  |  6,498.3 ns |   8.56 ns |   7.15 ns |  6,497.5 ns |  6,487.6 ns |  6,512.2 ns |  1.00 |    0.00 |      - |      - |      56 B |        1.00 |
|                                  |                      |            |                          |       |      |             |           |           |             |             |             |       |         |        |        |           |             |
| IterateForEach<String>           | ConcurrentDictionary | Job-DNQFPU | \CD-Baseline\corerun.exe | ?     | 512  | 11,520.5 ns |  15.80 ns |  12.34 ns | 11,522.8 ns | 11,495.6 ns | 11,535.1 ns |  1.00 |    0.00 |      - |      - |      64 B |        1.00 |
| IterateForEach<String>           | ConcurrentDictionary | Job-OSLZLB | \CD-Changed\corerun.exe  | ?     | 512  | 11,548.5 ns |  20.19 ns |  18.88 ns | 11,550.9 ns | 11,521.8 ns | 11,576.0 ns |  1.00 |    0.00 |      - |      - |      64 B |        1.00 |
|                                  |                      |            |                          |       |      |             |           |           |             |             |             |       |         |        |        |           |             |
| IndexerSet<Int32>                | ConcurrentDictionary | Job-DNQFPU | \CD-Baseline\corerun.exe | ?     | 512  | 12,158.4 ns |  24.32 ns |  21.56 ns | 12,158.5 ns | 12,116.4 ns | 12,192.8 ns |  1.00 |    0.00 |      - |      - |         - |          NA |
| IndexerSet<Int32>                | ConcurrentDictionary | Job-OSLZLB | \CD-Changed\corerun.exe  | ?     | 512  | 12,166.4 ns |  19.15 ns |  15.99 ns | 12,161.2 ns | 12,146.6 ns | 12,193.6 ns |  1.00 |    0.00 |      - |      - |         - |          NA |
|                                  |                      |            |                          |       |      |             |           |           |             |             |             |       |         |        |        |           |             |
| IndexerSet<String>               | ConcurrentDictionary | Job-DNQFPU | \CD-Baseline\corerun.exe | ?     | 512  | 18,730.4 ns |  49.19 ns |  43.60 ns | 18,721.9 ns | 18,679.0 ns | 18,815.9 ns |  1.00 |    0.00 |      - |      - |         - |          NA |
| IndexerSet<String>               | ConcurrentDictionary | Job-OSLZLB | \CD-Changed\corerun.exe  | ?     | 512  | 18,842.2 ns |  33.12 ns |  29.36 ns | 18,841.2 ns | 18,793.3 ns | 18,897.3 ns |  1.01 |    0.00 |      - |      - |         - |          NA |

// * Warnings *
Environment
  Summary -> Benchmark was executed on the virtual machine with Hyper-V hypervisor. Virtualization can affect the measurement result.

// * Hints *
Outliers
  CtorDefaultSize<Int32>.ConcurrentDictionary: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\CD-Baseline\corerun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=15, WarmupCount=1           -> 2 outliers were removed (153.19 ns, 154.72 ns)
  CtorDefaultSize<Int32>.ConcurrentDictionary: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\CD-Changed\corerun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=15, WarmupCount=1            -> 1 outlier  was  removed (154.26 ns)
  CtorDefaultSize<String>.ConcurrentDictionary: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\CD-Baseline\corerun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=15, WarmupCount=1          -> 1 outlier  was  removed (171.77 ns)
  CtorDefaultSize<String>.ConcurrentDictionary: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\CD-Changed\corerun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=15, WarmupCount=1           -> 1 outlier  was  removed, 2 outliers were detected (158.46 ns, 168.12 ns)
  TryAddDefaultSize<Int32>.ConcurrentDictionary: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\CD-Baseline\corerun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=15, WarmupCount=1         -> 2 outliers were removed (37.54 us, 38.74 us)
  TryAddDefaultSize<Int32>.ConcurrentDictionary: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\CD-Changed\corerun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=15, WarmupCount=1          -> 1 outlier  was  removed (38.60 us)
  TryAddDefaultSize<String>.ConcurrentDictionary: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\CD-Baseline\corerun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=15, WarmupCount=1        -> 1 outlier  was  removed (48.82 us)
  TryAddDefaultSize<String>.ConcurrentDictionary: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\CD-Changed\corerun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=15, WarmupCount=1         -> 1 outlier  was  removed (49.66 us)
  TryAddGiventSize<Int32>.ConcurrentDictionary: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\CD-Baseline\corerun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=15, WarmupCount=1          -> 1 outlier  was  removed (24.15 us)
  TryAddGiventSize<Int32>.ConcurrentDictionary: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\CD-Changed\corerun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=15, WarmupCount=1           -> 2 outliers were removed (24.12 us, 24.43 us)
  TryAddGiventSize<String>.ConcurrentDictionary: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\CD-Baseline\corerun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=15, WarmupCount=1         -> 1 outlier  was  removed (34.81 us)
  TryAddGiventSize<String>.ConcurrentDictionary: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\CD-Changed\corerun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=15, WarmupCount=1          -> 2 outliers were removed (34.63 us, 34.64 us)
  ContainsKeyFalse<Int32, Int32>.ConcurrentDictionary: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\CD-Baseline\corerun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=15, WarmupCount=1   -> 2 outliers were removed (2.18 us, 2.19 us)
  ContainsKeyFalse<String, String>.ConcurrentDictionary: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\CD-Baseline\corerun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=15, WarmupCount=1 -> 1 outlier  was  removed, 2 outliers were detected (6.64 us, 6.89 us)
  ContainsKeyFalse<String, String>.ConcurrentDictionary: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\CD-Changed\corerun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=15, WarmupCount=1  -> 2 outliers were removed (6.78 us, 6.82 us)
  ContainsKeyTrue<Int32, Int32>.ConcurrentDictionary: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\CD-Changed\corerun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=15, WarmupCount=1     -> 1 outlier  was  removed (2.36 us)
  ContainsKeyTrue<String, String>.ConcurrentDictionary: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\CD-Baseline\corerun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=15, WarmupCount=1  -> 1 outlier  was  removed (9.19 us)
  ContainsKeyTrue<String, String>.ConcurrentDictionary: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\CD-Changed\corerun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=15, WarmupCount=1   -> 2 outliers were removed, 3 outliers were detected (9.01 us, 9.15 us, 9.26 us)
  TryGetValueFalse<Int32, Int32>.ConcurrentDictionary: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\CD-Baseline\corerun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=15, WarmupCount=1   -> 2 outliers were removed (817.98 ns, 833.10 ns)
  TryGetValueFalse<Int32, Int32>.ConcurrentDictionary: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\CD-Changed\corerun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=15, WarmupCount=1    -> 2 outliers were removed (813.14 ns, 816.91 ns)
  TryGetValueFalse<String, String>.ConcurrentDictionary: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\CD-Baseline\corerun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=15, WarmupCount=1 -> 1 outlier  was  detected (5.35 us)
  TryGetValueFalse<String, String>.ConcurrentDictionary: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\CD-Changed\corerun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=15, WarmupCount=1  -> 2 outliers were removed (5.45 us, 5.53 us)
  TryGetValueTrue<Int32, Int32>.ConcurrentDictionary: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\CD-Baseline\corerun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=15, WarmupCount=1    -> 1 outlier  was  removed (1.05 us)
  TryGetValueTrue<String, String>.ConcurrentDictionary: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\CD-Baseline\corerun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=15, WarmupCount=1  -> 2 outliers were removed (6.97 us, 7.25 us)
  TryGetValueTrue<String, String>.ConcurrentDictionary: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\CD-Changed\corerun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=15, WarmupCount=1   -> 1 outlier  was  removed (6.76 us)
  CtorGivenSize<Int32>.ConcurrentDictionary: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\CD-Baseline\corerun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=15, WarmupCount=1             -> 1 outlier  was  removed, 3 outliers were detected (185.44 ns, 186.09 ns, 228.81 ns)
  CtorGivenSize<String>.ConcurrentDictionary: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\CD-Changed\corerun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=15, WarmupCount=1             -> 1 outlier  was  removed, 3 outliers were detected (198.83 ns, 201.65 ns, 213.77 ns)
  AddGivenSize<Int32>.ConcurrentDictionary: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\CD-Baseline\corerun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=15, WarmupCount=1              -> 2 outliers were removed (25.29 us, 25.42 us)
  AddGivenSize<Int32>.ConcurrentDictionary: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\CD-Changed\corerun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=15, WarmupCount=1               -> 1 outlier  was  removed (24.41 us)
  CtorFromCollection<Int32>.ConcurrentDictionary: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\CD-Baseline\corerun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=15, WarmupCount=1        -> 1 outlier  was  removed (21.82 us)
  CtorFromCollection<Int32>.ConcurrentDictionary: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\CD-Changed\corerun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=15, WarmupCount=1         -> 1 outlier  was  removed (21.64 us)
  CtorFromCollection<String>.ConcurrentDictionary: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\CD-Baseline\corerun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=15, WarmupCount=1       -> 1 outlier  was  removed (34.43 us)
  CtorFromCollection<String>.ConcurrentDictionary: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\CD-Changed\corerun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=15, WarmupCount=1        -> 2 outliers were removed (34.71 us, 42.85 us)
  CreateAddAndClear<Int32>.ConcurrentDictionary: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\CD-Baseline\corerun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=15, WarmupCount=1         -> 1 outlier  was  removed (43.18 us)
  CreateAddAndClear<Int32>.ConcurrentDictionary: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\CD-Changed\corerun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=15, WarmupCount=1          -> 2 outliers were removed (42.11 us, 42.93 us)
  CreateAddAndClear<String>.ConcurrentDictionary: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\CD-Baseline\corerun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=15, WarmupCount=1        -> 2 outliers were removed (53.98 us, 54.21 us)
  CreateAddAndClear<String>.ConcurrentDictionary: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\CD-Changed\corerun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=15, WarmupCount=1         -> 3 outliers were removed (54.38 us..58.26 us)
  IterateForEach<Int32>.ConcurrentDictionary: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\CD-Baseline\corerun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=15, WarmupCount=1            -> 1 outlier  was  removed (6.64 us)
  IterateForEach<Int32>.ConcurrentDictionary: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\CD-Changed\corerun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=15, WarmupCount=1             -> 2 outliers were removed (6.53 us, 6.56 us)
  IterateForEach<String>.ConcurrentDictionary: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\CD-Baseline\corerun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=15, WarmupCount=1           -> 3 outliers were removed, 4 outliers were detected (11.50 us, 11.57 us..11.66 us)
  IndexerSet<Int32>.ConcurrentDictionary: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\CD-Baseline\corerun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=15, WarmupCount=1                -> 1 outlier  was  removed (12.36 us)
  IndexerSet<Int32>.ConcurrentDictionary: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\CD-Changed\corerun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=15, WarmupCount=1                 -> 2 outliers were removed (12.24 us, 12.36 us)
  IndexerSet<String>.ConcurrentDictionary: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\CD-Baseline\corerun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=15, WarmupCount=1               -> 1 outlier  was  removed (19.00 us)
  IndexerSet<String>.ConcurrentDictionary: PowerPlanMode=00000000-0000-0000-0000-000000000000, Toolchain=\CD-Changed\corerun.exe, IterationTime=250ms, MaxIterationCount=20, MinIterationCount=15, WarmupCount=1                -> 1 outlier  was  removed (18.97 us)

// * Legends *
  Count       : Value of the 'Count' parameter
  Size        : Value of the 'Size' parameter
  Mean        : Arithmetic mean of all measurements
  Error       : Half of 99.9% confidence interval
  StdDev      : Standard deviation of all measurements
  Median      : Value separating the higher half of all measurements (50th percentile)
  Min         : Minimum
  Max         : Maximum
  Ratio       : Mean of the ratio distribution ([Current]/[Baseline])
  RatioSD     : Standard deviation of the ratio distribution ([Current]/[Baseline])
  Gen0        : GC Generation 0 collects per 1000 operations
  Gen1        : GC Generation 1 collects per 1000 operations
  Allocated   : Allocated memory per single operation (managed only, inclusive, 1KB = 1024B)
  Alloc Ratio : Allocated memory ratio distribution ([Current]/[Baseline])
  1 ns        : 1 Nanosecond (0.000000001 sec)
```

The following test case was also used to test the average number of accesses per unit of time the ConcurrentDictionary can handle with this change:

```C#
using System.Collections.Concurrent;
using System.Diagnostics;

const int KeyPoolCount = 128 << 10;
const int KeyPoolMask = KeyPoolCount - 1;

const int KeyCount = 4 << 10;
const int KeyMask = KeyCount - 1;

// Populate the key pool with unique nonzero keys. Zero is reserved for ownership.
var keyPool = new int[KeyPoolCount];
for (int i = 0; i < KeyPoolCount; i++)
    keyPool[i] = i + 1;

// Populate the keys that will exist in the CD with unique random keys from the pool. The pool entries from which the
// keys are taken are replaced with zeroes.
int rngSeed;
if (args.Length == 0)
{
    rngSeed = new Random().Next();
    Console.WriteLine($"Rng seed: {rngSeed}");
}
else
    rngSeed = (int)uint.Parse(args[0]);
var rng = new Random(rngSeed);
var keys = new int[KeyCount];
for (int i = 0; i < KeyCount; i++)
{
    int keyPoolIndex = rng.Next() & KeyPoolMask;
    int key = keyPool[keyPoolIndex];
    if (key == 0)
    {
        for (int j = keyPoolIndex + 1; ; j++)
        {
            keyPoolIndex = j & KeyPoolMask;
            key = keyPool[keyPoolIndex];
            if (key != 0)
                break;
        }
    }

    keyPool[keyPoolIndex] = 0;
    keys[i] = key;
}

// Populate the CD with the unique keys
var cd = new ConcurrentDictionary<int, int>();
for (int i = 0; i < KeyCount; i++)
{
    if (!cd.TryAdd(keys[i], 0))
        throw new InvalidOperationException();
}

int threadCount = Environment.ProcessorCount;
var threadIterationCounters = new int[(threadCount + 1) * 32];
ParameterizedThreadStart threadMain = data =>
{
    var tupleData = (Tuple<int, int>)data;
    int threadIndex = tupleData.Item1;
    int rngSeed = tupleData.Item2;
    var rng = new Random(rngSeed);
    ref int threadIterationCounter = ref threadIterationCounters[(threadIndex + 1) * 32];
    var localKeyPool = keyPool;
    var localKeys = keys;
    var localCd = cd;
    while (true)
    {
        int randomInt = rng.Next();

        // Select a new key and take ownership of the key pool entry
        int keyPoolIndex = randomInt & KeyPoolMask;
        int newKey = Interlocked.Exchange(ref localKeyPool[keyPoolIndex], 0);
        if (newKey == 0)
        {
            for (int j = keyPoolIndex + 1; ; j++)
            {
                keyPoolIndex = j & KeyPoolMask;
                newKey = Interlocked.Exchange(ref localKeyPool[keyPoolIndex], 0);
                if (newKey != 0)
                    break;
            }
        }

        // Select an existing key and take ownership of the existing key entry
        int keyIndex = randomInt & KeyMask;
        int key = Interlocked.Exchange(ref localKeys[keyIndex], 0);
        if (key == 0)
        {
            for (int j = keyIndex + 1; ; j++)
            {
                keyIndex = j & KeyMask;
                key = Interlocked.Exchange(ref localKeys[keyIndex], 0);
                if (key != 0)
                    break;
            }
        }

        if (!cd.TryRemove(key, out _) || !cd.TryAdd(newKey, 0))
            throw new InvalidOperationException();

        // Release the keys and ownerships
        localKeyPool[keyPoolIndex] = key;
        localKeys[keyIndex] = newKey;
        threadIterationCounter++;
    }
};

for (int i = 0; i < threadCount; i++)
{
    var t = new Thread(threadMain);
    t.IsBackground = true;
    t.Start(new Tuple<int, int>(i, rng.Next()));
}

var sw = new Stopwatch();

// Warmup
Thread.Sleep(6000);

// Measurements
int totalIterations = 0;
double totalDurationMs = 0;
for (int i = 0; i < 8; i++)
{
    int startIterations = 0;
    int endIterations = 0;

    for (int j = 0; j < threadCount; j++)
        startIterations += threadIterationCounters[(j + 1) * 32];
    sw.Restart();
    Thread.Sleep(1000);
    for (int j = 0; j < threadCount; j++)
        endIterations += threadIterationCounters[(j + 1) * 32];
    sw.Stop();

    int iterations = endIterations - startIterations;
    double durationMs = sw.Elapsed.TotalMilliseconds;
    Console.WriteLine($"Iterations/ms: {iterations / durationMs,10:0.0}");
    totalIterations += iterations;
    totalDurationMs += durationMs;
}

Console.WriteLine($"Average iter/ms: {totalIterations / totalDurationMs,10:0.0}");
```

After running the test several times, the results were the following:
|  | Baseline | Changed | Diff (%) |
|----------|----------|----------|------------|
|   Average iterations/ms  |   39528.2  |   39977.4 |    1.14%   |

The tests so far show that the performance of ConcurrentDictionary with these changes is relatively fine.

Reply:
The `CtorDefaultSize<Int32>` shows exactly the same allocation before/after. How is that possible? The new version is obviously allocating one more Lock object than it was before. The Lock objects are also bigger than the System.Objects that were being used before. How should I think about this?

Reply:
My bad, seems like I had some error while collecting the data, I updated the comment above